tinygrad/tensor.py

```diff
@@ -1317,6 +1317,24 @@ def unflatten(self, dim:int, sizes:Tuple[int,...]):
     dim = self._resolve_dim(dim)
     return self.reshape(self.shape[:dim] + sizes + self.shape[dim+1:])
 
+  def roll(self, shifts:Union[int, Tuple[int, ...]], dims:Union[int, Tuple[int, ...]]) -> Tensor:
+    """
+    Roll the tensor along specified dimension(s).
+    The rolling operation is circular, meaning that elements that go beyond the edge are wrapped around to the beginning of the dimension.
+
+    ```python exec="true" source="above" session="tensor" result="python"
+    print(Tensor.rand(3, 4, 1).roll(shifts=1, dims=0))
+    ```
+    ```python exec="true" source="above" session="tensor" result="python"
+    print(Tensor.rand(3, 4, 1).roll(shifts=-1, dims=0))
+    ```
+    """
+    dims, shifts = (dims,) if isinstance(dims, int) else dims, (shifts,) if isinstance(shifts, int) else shifts
+    dims = tuple(i % len(self.shape) for i in dims)
+    all_shifts = [shifts[dims.index(i)] % self.shape[i] if i in dims else 0 for i in range(len(self.shape))]
+    return functools.reduce(lambda t, args: t[tuple(slice(None) if j != args[0] else slice(-args[1], None) for j in range(len(t.shape)))].cat(
```

Comment:
Unreadable

Reply:
Is more readable now and tried to keep the lines minimal